files/common/usr/bin/illumos-cleanup

```diff
@@ -0,0 +1,185 @@
+#!/bin/bash
+
+POST_MIGRATION_CLEANUP=/var/delphix/migration/post-migration-cleanup-required
+FINAL_ILLUMOS_CLEANUP=/var/delphix/migration/final-illumos-cleanup-required
+
+function die() {
+	echo "$(basename "$0"): $*" >&2
+	exit 1
+}
+
+function echo_err() {
+	echo "$(basename "$0"): $*" >&2
+}
+
+function usage() {
+	echo "$(basename "$0"): $*" >&2
+	echo "Usage: $(basename "$0") post-migration"
+	echo "       $(basename "$0") final"
+	exit 2
+}
+
+function destroy_dataset() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy "$dataset"; then
+			echo "Destroyed '$dataset'"
+		else
+			echo_err "Failed to destroy '$dataset'"
+			return 1
+		fi
+	fi
+}
+
+function destroy_dataset_recursive() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy -r "$dataset"; then
+			echo "Destroyed '$dataset' recursively"
+		else
+			echo_err "Failed to destroy '$dataset' recursively"
+			return 1
+		fi
+	fi
+}
+
+function promote_dataset() {
+	local dataset="$1"
+	local origin
+
+	#
+	# To promote a dataset all the way to the top, zfs promote may need
+	# to be called multiple times if the dataset is a nested clone.
+	# In practice though, since we delete old os versions, the clone
+	# depth shouldn't go deeper than 2 or 3 levels.
+	#
+	for _ in {1..5}; do
```

Comment:
the comment says the limit is 2 or 3, so why do we use a limit of 5 here? should these be consistent with each other?

Reply:
I'll update the comment to mention that we want to stay on the safe side.